km/km_hcalls.c

```diff
@@ -322,6 +322,126 @@ static km_hc_ret_t sendrecvmsg_hcall(void* vcpu, int hc, km_hc_args_t* arg)
    return HC_CONTINUE;
 }
 
+/*
+ * Copy a 'struct msghdr' from <sys/socket.h from guest space to km space.
+ * We need to do this becuase the msghdr passes iovec's between user land and
+ * kernel land and they contain addresses that need translation.
+ */
+static int copyin_msghdr(struct msghdr* km_msg, struct msghdr* guest_msg)
+{
+   if (guest_msg->msg_name != NULL) {
+      km_msg->msg_name = km_gva_to_kma((uintptr_t)guest_msg->msg_name);
+      if (km_msg->msg_name == NULL) {
+         return -EFAULT;
+      }
+   } else {
+      km_msg->msg_name = NULL;
+   }
+   km_msg->msg_namelen = guest_msg->msg_namelen;
+   if (guest_msg->msg_iovlen >= UIO_MAXIOV) {
+      return -EINVAL;
+   }
+
+   if (km_msg->msg_iovlen > 0) {
+      for (int i = 0; i < km_msg->msg_iovlen; i++) {
+         // Don't validate addresses if the element has a length of zero.
+         if (guest_msg->msg_iov[i].iov_len > 0) {
+            km_msg->msg_iov[i].iov_len = guest_msg->msg_iov[i].iov_len;
+            km_msg->msg_iov[i].iov_base = km_gva_to_kma((uintptr_t)guest_msg->msg_iov[i].iov_base);
+            if (km_msg->msg_iov[i].iov_base == NULL) {
+               return -EFAULT;
+            }
+         }
+      }
+   }
+   if (guest_msg->msg_control != NULL) {
+      km_msg->msg_control = km_gva_to_kma((uintptr_t)guest_msg->msg_control);
+      if (km_msg->msg_control == NULL) {
+         return -EFAULT;
+      }
+   }
+   km_msg->msg_controllen = guest_msg->msg_controllen;
+   km_msg->msg_flags = guest_msg->msg_flags;
+   return 0;
+}
+
+static km_hc_ret_t sendrecvmmsg_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   // int sendmmsg(int sockfd, struct mmsghdr *msgvec, unsigned int vlen, int flags);
+   // int recvmmsg(int sockfd, struct mmsghdr *msgvec, unsigned int vlen, int flags, struct timespec
+   // *timeout);
+
+   if (arg->arg2 == 0) {
+      arg->hc_ret = -EFAULT;
+      return HC_CONTINUE;
+   }
+   struct mmsghdr* guest_mmsghdr = km_gva_to_kma(arg->arg2);
+
+   // Can't have more then UIO_MAXIOV mmsghdr's
+   if (arg->arg3 == 0 || arg->arg3 >= UIO_MAXIOV) {
```

Comment:
Is vlen limited to UIO_MAXIOV - 1?
The man page says:
"The value specified in vlen is capped to UIO_MAXIOV (1024)."
Kind of feels like UIO_MAXIOV is the largest allowed value.

From kernel source in net/socket.c:
In __sys_sendmmsg()
```
        if (vlen > UIO_MAXIOV)
                vlen = UIO_MAXIOV;
```
It's not even a failure, it just silently drops the excess.

Reply:
Yep. You're right. Changed to match kernel behavior.